### PR TITLE
Add HttpConnectionsHealthCheck

### LINKED ISSUE #28: Add HttpConnectionsHealthCheck

### PR DESCRIPTION
**Pull request by chrisrohr**
Closes #28

### CODE REVIEW

File: src/main/java/org/kiwiproject/dropwizard/util/health/HttpConnectionsHealthCheck.java [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+import java.util.Set;
+import java.util.SortedMap;
+
+/**
+ * Health check that checks the percent of leased connections against the maximum number of connections for
+ * JAX-RS {@link javax.ws.rs.client.Client} instances that were created using Dropwizard's
+ * {@code io.dropwizard.client.JerseyClientBuilder}, which creates an HTTP connection pool and registers various
+ * connection metrics that we can query.
+ * <p>
+ * Please note, if using a default JAX-RS {@link javax.ws.rs.client.Client} created using the normal
```
Comment (sleberknight) [inline src/main/java/org/kiwiproject/dropwizard/util/health/HttpConnectionsHealthCheck.java:32]: This comment made me go back into registry-aware-jersey-client...the regular `ClientBuilders.jersey()` also returns a `Client` that doesn't have any metrics attached. We do provide the `DropwizardManagedClientBuilder` which uses the Dropwizard stuff and attaches metrics. Do we care that clients built using `ClientBuilders.jersey()` don't have metrics? I guess they also don't have an HTTP connection pool, either...

File: src/main/java/org/kiwiproject/dropwizard/util/health/HttpConnectionsHealthCheck.java [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+                .warningThreshold(warningThreshold)
+                .build();
+
+        LOG.trace("{}: {} of {} leased ({}%)", clientName, leasedConnections, maxConnections,
+                connectionInfo.percentLeased);
+
+        return connectionInfo;
+    }
+
+    private Result determineResult(Map<ClientConnectionInfo.HealthStatus, List<ClientConnectionInfo>> clientHealth,
```
Comment (sleberknight) [inline src/main/java/org/kiwiproject/dropwizard/util/health/HttpConnectionsHealthCheck.java:125]: nit: move ` List<ClientConnectionInfo>> clientHealth` to separate line.

File: src/test/java/org/kiwiproject/dropwizard/util/health/HttpConnectionsHealthCheckTest.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    void setUp() {
+        var environment = DropwizardMockitoMocks.mockEnvironment();
+        metrics = DropwizardMockitoMocks.mockMetricRegistry(environment);
+        healthCheck = new HttpConnectionsHealthCheck(metrics);
+    }
+
+    @Test
+    void testClientIsNotBeingDumb_ByCheckingForInvalidThreshold(SoftAssertions softly) {
+        softly.assertThatThrownBy(() -> newHealthCheckWithThreshold(0.0)).isExactlyInstanceOf(IllegalArgumentException.class);
+        softly.assertThat(catchThrowable(() -> newHealthCheckWithThreshold(1.0))).isNull();
```
Comment (sleberknight) [inline src/test/java/org/kiwiproject/dropwizard/util/health/HttpConnectionsHealthCheckTest.java:43]: This is confusing....why not replace with `assertThatCode(...).doesNotThrowAnyException()` ?

Could put each of these as separate tests in a `@Nested` class

File: src/test/java/org/kiwiproject/dropwizard/util/health/HttpConnectionsHealthCheckTest.java [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        return new HttpConnectionsHealthCheck(metrics, value);
+    }
+
+    @Test
+    void testCheck_WhenNoDropwizardJerseyClients(SoftAssertions softly) {
+        when(metrics.getGauges(any())).thenReturn(new TreeMap<>());
+
+        HealthCheck.Result result = healthCheck.check();
+
+        softly.assertThat(result.isHealthy()).isTrue();
```
Comment (sleberknight) [inline src/test/java/org/kiwiproject/dropwizard/util/health/HttpConnectionsHealthCheckTest.java:59]: suggest changing these to use the fluent `HealthCheckResultAssertions` from kiwi-test...same with others that do the same assertions

e.g.

```java
assertThatHealthCheck(healthCheck)
    .isHealthy()
    .hasMessage("No HTTP clients found with metrics");
```

Yes, we lose the soft assertions, but I prefer the fluent aspect more.

And maybe we could (should?) change `HealthCheckResultAssertions` so it uses soft assertions internally...

**Review by sleberknight — CHANGES_REQUESTED** [review]

File: src/main/java/org/kiwiproject/dropwizard/util/health/HttpConnectionsHealthCheck.java (hunk 1 above)
Comment (chrisrohr) [reply src/main/java/org/kiwiproject/dropwizard/util/health/HttpConnectionsHealthCheck.java:32]: Yeah that is an issue, but hopefully we will be using the dropwizard managed ones.

File: src/main/java/org/kiwiproject/dropwizard/util/health/HttpConnectionsHealthCheck.java (hunk 2 above)
Comment (chrisrohr) [reply src/main/java/org/kiwiproject/dropwizard/util/health/HttpConnectionsHealthCheck.java:125]: Even though it is part of the generic Map definition?

Comment (sleberknight) [reply src/main/java/org/kiwiproject/dropwizard/util/health/HttpConnectionsHealthCheck.java:125]: Oops my bad. No.

File: src/test/java/org/kiwiproject/dropwizard/util/health/HttpConnectionsHealthCheckTest.java [hunk 5]
```diff
@@ -37,117 +37,222 @@ void setUp() {
...
+                        maxGaugeNameFor("some-client"), gaugeReturning(10)
+                );
+
+                when(metrics.getGauges(any())).thenReturn(gauges);
+
+                assertThatHealthCheck(healthCheck)
+                        .isHealthy()
+                        .hasMessage("1 HTTP client(s) < 50.0% leased connections.");
+
+                // NOTE: Yes this is calling the check method a second time, but I need the details map for
```
Comment (sleberknight) [inline src/test/java/org/kiwiproject/dropwizard/util/health/HttpConnectionsHealthCheckTest.java:91]: Instead, just use `assertThatResult`.

e.g.

```java
var result = assertThatResult(result)...

// custom assertions on result
```

File: src/test/java/org/kiwiproject/dropwizard/util/health/HttpConnectionsHealthCheckTest.java [hunk 6]
```diff
@@ -37,117 +37,222 @@ void setUp() {
...
+                        maxGaugeNameFor("another-client"), gaugeReturning(10)
+                );
+
+                when(metrics.getGauges(any())).thenReturn(gauges);
+
+                assertThatHealthCheck(healthCheck)
+                        .isHealthy()
+                        .hasMessage("2 HTTP client(s) < 50.0% leased connections.");
+
+                // NOTE: Yes this is calling the check method a second time, but I need the details map for
```
Comment (sleberknight) [inline src/test/java/org/kiwiproject/dropwizard/util/health/HttpConnectionsHealthCheckTest.java:129]: ditto here and all the ones below re: `assertThatResult`

**Review by sleberknight — APPROVED** [review]

**Pull request merged**